include/arch/riscv/thread.h

```diff
@@ -70,6 +110,19 @@ typedef struct _callee_saved _callee_saved_t;
 
 struct _thread_arch {
 	uint32_t swap_return_value; /* Return value of z_swap() */
+
+#ifdef CONFIG_RISCV_PMP
+	ulong_t s_pmpcfg[PMP_CFG_CSR_NUM_FOR_STACK_GUARD];
+	ulong_t s_pmpaddr[PMP_REGION_NUM_FOR_STACK_GUARD];
+#endif
+
+#ifdef CONFIG_USERSPACE
```

Comment:
These parameters are for userspace feature and block above are for stack guard. Both use PMP but are uncorrelated.

Reply:
They are correlated, one requires the other. 😁 I actually ran into a compiler error here when running one of the userspace tests because it enabled USERSPACE but I did not enable CONFIG_PMP.

Reply:
They are correlated, one requires the other. 😁 I actually ran into a compiler error here when running one of the userspace tests because it enabled USERSPACE but I did not enable CONFIG_PMP.

Reply:
- Userspace is correalted to PMP
- Stack Guard is correlated to PMP
- Userspace is UNCORRELATED to Stack guard

Reply:
Then I will not put Userspace specific parameters under RISCV_PMP because that will take them if you just enable stackguard

Reply:
I think you misunderstood my original comment. I was not suggesting we replace this with CONFIG_RISCV_PMP but also depend on CONFIG_RISCV_PMP. But the root of the issue is it shouldn't be possible to enable USERSPACE without RISCV_PMP under RISCV architectures. It currently is and the only indication is an unobvious compiler error here.

Reply:
Right I misunderstood sorry. Isn't worth to have compilation fail if you enable userspace without PMP flag ? If I fix here the #ifdef by adding "&& CONFIG_RISCV_PMP" the compilation should fail otherwhere. If you think fix it is worth, I will do it :)

Reply:
I think your right that it will fail elsewhere, maybe the fix is to modify Kconfig so this configuration is now allowed.

Reply:
In fact, the security is here in the Kconfig:
config CORE_E31
	bool "Use E31 core"
	select RISCV_PMP
	select ARCH_HAS_USERSPACE

Do you think ARCH_HAS_USERSPACE should be under RISCV_PMP config ? Is that make sense ?

Reply:
RISCV_PMP implies ARCH_HAS_USERSPACE for sure. ARM does something similar:

```
config CPU_CORTEX_M
        ...
        select ARCH_HAS_USERSPACE if ARM_MPU
        ...
```

Reply:
Ok but I can't do "select RISCV_PMP" and "select ARCH_HAS_USERSPACE if RISCV_PMP"
ARM must do 2 things: 
- Enable ARCH_HAS_MPU in SoC directory Kconfig file.
- Then you are able to enable manually ARM_MPU.

For RISCV SoC Kconfig seems not use ARCH_HAS_*** flags, and I don't know it it should handle this like ARM (even if technically it can).
My implementation avoid that.

It should be great if a RISCV contributor could give its point of view. And I'm not sure about which RISCV's Kconfig I have to add "select ARCH_HAS_MPU"  in soc/riscv/*

Reply:
@xodus7 I mean: I think my way is great (I think enough) but if you don't think that must not be merged like that, no worry, just let me know and I will fix it.

Reply:
I just ran into this while setting it up for my system and ran into these issues. It would be nice if we could make the misconfiguration more obvious but I don't think it's a blocker.